mite_selenium/__init__.py

```diff
@@ -51,32 +56,119 @@ def start(self):
-    """Maybe would probably need more error handling
-    if somebody tried to use with an incompatible browser"""
-    if context.browser.capabilities['browserName'] != 'chrome':
-        logger.warning(
-            'Timing metrics in %s are not currently supported',
-            context.browser.capabilities['browserName'],
-        )
-        return
-
-    timings = context.browser.execute_script(
-        'return performance.getEntriesByType("navigation")'
-    )[0]
-    context.send(
-        "http_selenium_metrics",
-        dom_complete=timings["domComplete"],
-        dom_interactive=timings["domInteractive"],
-        transfer_size=timings["transferSize"],
-        transfer_start_timings=["requestStart"],
-        tls_time=timings["secureConnectionStart"],
-        dns_time=timings["domainLookupEnd"],
-        connect_time=timings["connectEnd"],
-        total_time=timings["duration"],
-        effective_url=timings["name"],
-    )
+    def _browser_has_timing_capabilities(self):
+        return self._context.browser.capabilities["browserName"] != "chrome"
+
+    def _is_using_tls(self, name):
+        return name.startsWith("https")
+
+    def _get_tls_timing(self, timings):
+        if self._is_using_tls(timings["name"]):
+            return timings["connectEnd"] - timings["secureConnectionStart"]
+        else:
+            logger.warning("Secure TLS connection not used, defaulting tls_time to 0")
+            return 0
+
+    def _get_tcp_timing(self, timings):
+        if self._is_using_tls(timings["name"]):
+            return timings["secureConnectionStart"] - timings["connectStart"]
+        else:
+            return timings["connectEnd"] - timings["connectStart"]
+
+    def _send_page_load_metrics(self):
+        if self._browser_has_timing_capabilities():
+            timings, *_ = self._context.browser.execute_script(
+                "return performance.getEntriesByType('navigation')"
+            )
+            protocol = timings["nextHopProtocol"]
+            if protocol != "http/1.1":
+                logger.warning(
+                    f"Timings may be inaccurate as protocol is not http/1.1: {protocol}"
+                )
+            metrics = {
+                "tcp_time": self._get_tcp_timing(timings),
+                "tls_time": self._get_tls_time_if_applicable(timings),
+                "dns_lookup_time": timings["domainLookupEnd"]
+                - timings["domainLookupStart"],
+                "time_to_first_byte": timings["responseStart"] - timings["connectionEnd"],
+                "time_to_last_byte": timings["responseEnd"] - timings["connectionEnd"],
+                "page_weight": timings["transferSize"],
+            }
+            self._context.send(
+                "http_selenium_metrics",
+                *self._extract_and_convert_metrics_to_seconds(metrics),
+            )
+
+    def _extract_and_convert_metrics_to_seconds(self, metrics):
+        converted_metrics = defaultdict()
```

Comment:
```suggestion
        converted_metrics = dict()
```

Unless I'm missing something, defaultdict isn't needed here...

Reply:
Misunderstood the use case for defaultdict, changed this now